server/src/main/java/org/opensearch/threadpool/RunnableTaskListenerFactory.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.threadpool;
+
+import org.apache.lucene.util.SetOnce;
+
+import java.util.function.Function;
+
+public class RunnableTaskListenerFactory implements Function<RunnableTaskExecutionListener, RunnableTaskExecutionListener> {
+
+    private final SetOnce<RunnableTaskExecutionListener> listener = new SetOnce<>();
+
+    @Override
+    public RunnableTaskExecutionListener apply(RunnableTaskExecutionListener runnableTaskExecutionListener) {
+        listener.set(runnableTaskExecutionListener);
+        return listener.get();
+    }
+
+    public RunnableTaskExecutionListener get() {
+        assert listener.get() != null;
+        return listener.get();
+    }
+}
```

Comment:
This modelling is more of a Provider than a Factory. I would recommend creating a functional interface for provider and use AtomicReference to set it